### PR TITLE
Trustdomain

### PR DESCRIPTION
**Pull request by willscott**
Finish trustdomain / xoring of pad on server side.

### CODE REVIEW

File: libtalek/subscription.go [hunk 1]
```diff
@@ -74,13 +74,13 @@ func (s *Subscription) generatePoll(config *ClientConfig, seqNo uint64) (*common
 	// The first Trust domain is the one with the explicit bucket bit-flip.
 	k0, k1 = s.Seed2.KeyUint128()
 	bucket2 := siphash.Hash(k0, k1, seqNoBytes) % config.CommonConfig.NumBuckets
-	args[1].TD[0].RequestVector = make([]byte, config.CommonConfig.NumBuckets/8+1)
-	args[1].TD[0].RequestVector[bucket2/8] |= 1 << (bucket1 % 8)
+	args[1].TD[0].RequestVector = make([]byte, (config.CommonConfig.NumBuckets+7)/8)
+	args[1].TD[0].RequestVector[bucket2/8] |= 1 << (bucket2 % 8)
```
Comment (willscott) [inline libtalek/subscription.go]: This was a bug. this was the part copied from the args[0] which is for bucket1 copied down to now be for args[1] (bucket2), but one instance hadn't been changed yet to bucket2.

File: libtalek/client.go [hunk 2]
```diff
@@ -89,24 +88,8 @@ func (c *Client) Ping() bool {
...
-// TODO: support messages spanning multiple data items.
-func (c *Client) MaxLength() int {
-	config := c.config.Load().(ClientConfig)
-	return config.DataSize
-}
-
 func (c *Client) Publish(handle *Topic, data []byte) error {
 	config := c.config.Load().(ClientConfig)
-
-	if len(data) > config.DataSize {
```
Comment (ryscheng) [inline libtalek/client.go]: Nvm on this one. I see you added it back on a later PR

File: libtalek/client_test.go [hunk 3]
```diff
@@ -70,4 +76,49 @@ func TestWrite(t *testing.T) {
...
+	c := NewClient("TestClient", config, &leader)
+	if c == nil {
+		t.Fatalf("Error creating client")
+	}
+
+	handle, _ := NewTopic()
+
+	// Recreate the expected buckets to make sure we're seeing
+	// the real write.
+	var seqNoBytes [12]byte
```
Comment (willscott) [inline libtalek/client_test.go]: factored out a nextBucket method to reduce the complexity of re-generating the next buckets

**Comment by ryscheng** [comment]
Sure makes sense

**Comment by willscott** [comment]
added a logic test of that in the existing testPoll

File: libtalek/subscription.go [hunk 4]
```diff
@@ -37,22 +38,36 @@ func NewSubscription() (s *Subscription, err error) {
...
 func (s *Subscription) generatePoll(config *ClientConfig, seqNo uint64) (*common.ReadArgs, *common.ReadArgs, error) {
 	if s.SharedSecret == nil || s.SigningPublicKey == nil {
 		return nil, nil, errors.New("Subscription not fully initialized")
 	}
 
 	args := make([]*common.ReadArgs, 2)
-	seqNoBytes := make([]byte, 12)
-	_ = binary.PutUvarint(seqNoBytes, seqNo)
+	bucket1, bucket2 := s.nextBuckets()
+	bucket1 %= config.CommonConfig.NumBuckets
```
Comment (ryscheng) [inline libtalek/subscription.go]: Shouldn't this be included in the nextbuckets() ?

Comment (willscott) [inline libtalek/subscription.go]: Maybe? i'm not super happy with either option.

Right now, topics / handles don't have a specific talek config ingrained in them. So either, we would need a CommonConfig argument passed to the nextBuckets method, or have it be part of the topic.

writing this out, the simplest is probably to pass in config as an argument.

Comment (ryscheng) [inline libtalek/subscription.go]: I see your point. I didn't think of that.
I think you are right, adding a parameter of numBuckets is probably sufficient.
I don't think it's necessary to pass all of config right?

**Comment by willscott** [comment]
Change made. I'm going to take this as good and flatten outstanding changes into master.

**Pull request merged**